Cgi fix test14

edouardproust opened the pull request:
- Clarified the code in Network: variables or methods naming, etc
- Added wrappers _safeRecv and _safeSend including poll(), but AI says it is redondant... Could you pleas einvestigate if this is true?
- Added ealry check in Netwrok if request length is above ABSOLUTE_MAX_CLIENT_SIZE: if it is returns direclty a 413 response
- I move Network::isRequestComplete in RequestParser::isRequestComplete. And i refactored it as much as i could to make it more robust... But it is far from being perfect still... I didn't manage to reuse the methods you made in RequestParsing for doing this work

Please verify the points above and that nothing was borken in the process, thanks Ava

edouardproust committed:
before checkout

edouardproust committed:
before merge

edouardproust committed:
merge w/ main

edouardproust committed:
total change

edouardproust committed:
refactored Network

edouardproust committed:
before errno fix

edouardproust committed:
in progress

edouardproust committed:
made network module async

edouardproust committed:
before rollback on safeRecv and safeSend

Skoteini-42 commented:
Thoughts on Netowrk + requirements from subject + eval sheet :

- I think we're checking poll() mutliple times though `_safeRecv() ` and `_safeSend`, and I think we violate these lines from the subject if I undertstand correctly : "I/O that can wait for data (sockets, pipes/FIFOs, etc.) must be non-blocking **and driven by a single poll() (or equivalent)**. Calling read/recv or write/send on these descriptors without prior readiness will result in a grade of 0."

Remember when we tralked about how this works like threads do? You said that you had to create wrappers for send() and recv() in a way that mutexes were for philosophers - and I totally understood that , but I think if we're using poll() inside those methods we violate that requirement. Maybe they could serve the same purpose but without those poll() checks you think ?

But if I am correct about this, I think you've fixed all the other issues that we had before, and looks like a great job!
And I did several tests with raw requests, they seem to be being read correctly. I also checked the connections (keep-alive and close) and they still work well.

I also ran siege tests again, and it seems we have 100% efficiency on the tests I did, so that's also a very good sign that the refactored Network is working well.

edouardproust committed:
fixed tests/website/cgi/test.py + tests/website/errors/404.htm

edouardproust committed:
simplified _disconnectCLient and _prepareClientForNextRequestl

edouardproust committed:
refactored

edouardproust committed:
started adding progressive response send

edouardproust committed:
made non-blocking send by chunks

edouardproust commented:
- Ok i removed the wrappers for revc and send. 
- I refactored the response sending so it is now non-blocking (sending by chunks)
- I renammed even better methods and reorganized even more the code, i hope it is clear like that
- I added comments to explain the role of each method and added @notes in method comments to justify that the code follows the subjet's requirements (please check that it is actually correct, i may have missed something... thanks)

---

How to test:
- you can run `make test_dev` and navigate in the links, run curl tests
- then do `make test_42` and in another tarminal `./tests/ubuntu_tester http://localhost:8080`.
- When it is sending the big 100MB request, you can visit in the browser `http://localhost:8080`, then check the logs to verify that the server continue collectiong the 100MB request (not blocked by the display of http://localhost:8080 in the browser)

---

TODO:

- Now i have to implement de CGI in a non blocking manner in this new Network logic (some comments in main loop show where)

Skoteini-42 reviewed (approved):
Excuse my words but fucking excellent work!
Tested with siege, tested again keep-alive/close connections - all good!
I can see no violations anymore concerning the requirements, GOOD JOB!
